contracts/bid/src/hosted/transaction.rs

```diff
@@ -0,0 +1,286 @@
+    /// contract storage which will return some gas to the caller.
+    pub fn withdraw(
+        &mut self,
+        sig: Signature,
+        pk: PublicKey,
+        _note: Note,
+        _spend_proof: Vec<u8>,
+        block_height: u64,
+    ) -> bool {
+        // Setup error flag to false
+        let mut err_flag = false;
+        // Check wether there's an entry on the map for the pk.
+        let idx = match self.key_idx_map().get(pk) {
+            // If no entries are found for this PK it's just an err since there
+            // are no bids related to this PK to be extended.
+            Ok(None) => {
+                err_flag = true;
+                usize::MAX
+            }
+            Ok(Some(idx)) => *idx as usize,
+            Err(_) => {
+                err_flag = true;
+                usize::MAX
+            }
+        };
+
+        // In case there was an error, we simply return
+        if err_flag && idx == usize::MAX {
+            return err_flag;
+        }
+
+        // Fetch bid info from the tree. Note that we can safely unwrap here due
+        // to the checks done previously while getting the idx from the map.
+        let bid = self
+            .tree()
+            .get(idx as u64)
+            .expect("Unexpected error. Map & Tree are out of sync.");
+
+        if bid.0.expiration() < (block_height + COOLDOWN_PERIOD) {
+            // If we arrived here, the bid is elegible for withdrawal.
+            // Now we need to check wether the signature is correct.
+            // Verify schnorr sig.
+            if !fake_abi::verify_schnorr_sig(
+                pk,
+                sig,
+                BlsScalar::from(bid.0.expiration()),
+            ) {
+                err_flag = true;
+                return err_flag;
+            };
+            // Inter contract call
+
+            // If the inter-contract call succeeds, we need to clean the
+            // tree & map. Note that if we clean the entry
+            // corresponding to this `PublicKey` from the
+            // map there will be no need to do so from the tree. Since the
+            // rest of the functions rely on the map to gain
+            // access to the bid that is inside of the tree.
+            self.key_idx_map_mut()
+                .remove(pk)
+                .expect("Canon Store error happened.");
+            // TODO: Zeroize in the tree the leaf that corresponds to the idx
+            // linked to `pk` in the map.
+            // See: https://github.com/dusk-network/rusk/issues/164
+            err_flag
+        } else {
+            err_flag = true;
+            err_flag
+        }
+    }
+}
+
+// TODO: Until PLONK is no_std compatible and we can serialize PublicInputs
+use dusk_bytes::Serializable;
+use dusk_jubjub::JubJubScalar;
+
+const BLS_SCALAR: u8 = 1;
+const JUBJUB_SCALAR: u8 = 2;
+const JUBJUB_AFFINE: u8 = 3;
+
+/// Public Input
+#[derive(Debug, Copy, Clone)]
+enum PublicInput {
```

Comment:
Redundant structure redefinition. AFAIK these should be temporary so we should have an associated tech-debt with a TODO

Reply:
There's a comment on the top.